File: akkaserverpersistent/src/main/java/com/lightbend/java/modelserver/actor/presistence/FilePersistence.java
package com.lightbend.java.modelserver.actor.presistence;

import com.lightbend.model.Model;
import com.lightbend.model.ModelFactory;
import com.lightbend.model.ModelServingInfo;
import com.lightbend.model.Modeldescriptor;
import com.lightbend.model.PMML.PMMLModelFactory;
import com.lightbend.model.tensorflow.TensorflowModelFactory;

import java.io.*;
import java.util.HashMap;
import java.util.Map;
import java.util.Optional;

public class FilePersistence {

    private final static String  basDir = "persistence";
    private static final Map<Integer, ModelFactory> factories = new HashMap<Integer, ModelFactory>() {
        {
            put(Modeldescriptor.ModelDescriptor.ModelType.TENSORFLOW.getNumber(), TensorflowModelFactory.getInstance());
            put(Modeldescriptor.ModelDescriptor.ModelType.PMML.getNumber(), PMMLModelFactory.getInstance());
        }
    };

    private FilePersistence(){}

    public static ExecutionState restoreState(String  dataType){
        Optional<DataInputStream> mayBeStream = getDataInputStream(dataType);
        if(mayBeStream.isPresent())
            return new ExecutionState(readModel(mayBeStream.get()), readServingInfo(mayBeStream.get()));
        else
            return new ExecutionState();
    }

    private static Optional<DataInputStream> getDataInputStream(String fileName) {
        File file = new File(basDir + "/" + fileName);
        if(file.exists())
            try {
                return Optional.of(new DataInputStream(new FileInputStream(file)));
            }catch (Throwable t){
                System.out.println("Error Deserializing model");
                t.printStackTrace();
                return Optional.empty();
            }
        else
            return Optional.empty();
    }

    private static Optional<Model> readModel(DataInputStream input) {

        try {
            int length = (int)input.readLong();
            if (length == 0)
                return Optional.empty();
            int type = (int) input.readLong();
            byte[] bytes = new byte[length];
            input.read(bytes);
            ModelFactory factory = factories.get(type);
            return Optional.of(factory.restore(bytes));
        } catch (Throwable t) {
            System.out.println("Error Deserializing model");
            t.printStackTrace();
            return Optional.empty();
        }
    }

    private static Optional<ModelServingInfo> readServingInfo(DataInputStream input){
        try {
            long length = input.readLong();
            if (length == 0)
                return null;
            String descriprtion = input.readUTF();
            String name = input.readUTF();
            double duration = input.readDouble();
            long invocations = input.readLong();
            long max  = input.readLong();
            long min = input.readLong();
            long since = input.readLong();
            return Optional.of(new ModelServingInfo(name, descriprtion, since, invocations, duration, min, max));
        } catch (Throwable t) {
            System.out.println("Error Deserializing serving info");
            t.printStackTrace();
            return Optional.empty();
        }
    }

    public static void saveState(String dataType, Model model, ModelServingInfo servingInfo){
        Optional<DataOutputStream> mayBeOutput = getDataOutputStream(dataType);
        if(mayBeOutput.isPresent()) {
            writeModel(model, mayBeOutput.get());
            writeServingInfo(servingInfo, mayBeOutput.get());
            try {
                mayBeOutput.get().flush();
                mayBeOutput.get().close();
            }catch (Throwable t){
                System.out.println("Error Deserializing serving info");
                t.printStackTrace();
            }
        }
    }

    private static Optional<DataOutputStream> getDataOutputStream(String fileName){
        try {
            File dir = new File(basDir);
            if (!dir.exists()) {
                dir.mkdir();
            }
            File file = new File(dir, fileName);
            if (!file.exists())
                file.createNewFile();
            return Optional.of(new DataOutputStream(new FileOutputStream(file)));
        }
        catch (Throwable t){
            System.out.println("Error Deserializing serving info");
            t.printStackTrace();
            return Optional.empty();
        }
    }

    private static void writeModel(Model model, DataOutputStream output){
        try{
            if(model == null){
                output.writeLong(0);
                return;
            }
            byte[] bytes = model.getBytes();
            output.writeLong(bytes.length);
            output.writeLong(model.getType());
            output.write(bytes);
        }
        catch (Throwable t){
            System.out.println("Error Serializing model");
            t.printStackTrace();
        }
    }

    private static void writeServingInfo(ModelServingInfo servingInfo, DataOutputStream output){
        try{
            if(servingInfo == null) {
                output.writeLong(0);
                return;
            }
            output.writeLong(5);
            output.writeUTF(servingInfo.getDescription());
            output.writeUTF(servingInfo.getName());
            output.writeDouble(servingInfo.getDuration());
            output.writeLong(servingInfo.getInvocations());
            output.writeLong(servingInfo.getMax());
            output.writeLong(servingInfo.getMin());
            output.writeLong(servingInfo.getSince());
        }
        catch (Throwable t){
            System.out.println("Error Serializing servingInfo");
            t.printStackTrace();
        }
    }
}

File: akkaserverpersistent/src/main/scala/com/lightbend/modelserver/actor/persistence/FilePersistence.scala
package com.lightbend.modelserver.actor.persistence


import java.io.{DataInputStream, DataOutputStream, FileInputStream, FileOutputStream, File}

import com.lightbend.model.modeldescriptor.ModelDescriptor
import com.lightbend.modelServer.model.{Model, ModelToServeStats}
import com.lightbend.modelServer.model.PMML.PMMLModel
import com.lightbend.modelServer.model.tensorflow.TensorFlowModel

object FilePersistence {

  private final val basDir = "persistence"

  private val factories = Map(
    ModelDescriptor.ModelType.PMML.index -> PMMLModel,
    ModelDescriptor.ModelType.TENSORFLOW.index -> TensorFlowModel
  )

  def restoreState(dataType: String) : (Option[Model], Option[ModelToServeStats]) = {
    getDataInputStream(dataType) match {
      case Some(input) => (readModel(input), readServingInfo(input))
      case _ => (None, None)
    }
   }

  private def getDataInputStream(fileName: String) : Option[DataInputStream] = {
    val file = new File(basDir + "/" + fileName)
    file.exists() match {
      case true => Some(new DataInputStream(new FileInputStream(file)))
      case _ => None
    }
  }

  private def readModel(input : DataInputStream) : Option[Model] = {
    input.readLong.toInt match{
      case length if length > 0 => {
        val `type` = input.readLong.toInt
        val bytes = new Array[Byte](length)
        input.read(bytes)
        factories.get(`type`) match{
          case Some(factory) => try {
            Some(factory.restore(bytes))
          }
          catch {
            case t: Throwable =>
              System.out.println("Error Deserializing model")
              t.printStackTrace()
              None
          }
          case _ => None
        }
      }
      case _ => None
    }
  }

  private def readServingInfo(input: DataInputStream) : Option[ModelToServeStats] = {
    input.readLong match {
      case length if length > 0 => {
        try {
          Some(ModelToServeStats(input.readUTF, input.readUTF, input.readLong, input.readLong, input.readDouble, input.readLong, input.readLong))
        } catch {
          case t: Throwable =>
            System.out.println("Error Deserializing serving info")
            t.printStackTrace()
            None
        }
      }
      case _ => None
    }
  }


  def saveState(dataType: String, model: Model, servingInfo: ModelToServeStats) : Unit = {
    val output = getDataOutputStream(dataType)
    writeModel(output, model)
    writeServingInfo(output, servingInfo)
    output.flush()
    output.close()
  }

  private def getDataOutputStream(fileName: String) : DataOutputStream = {

    val dir = new File(basDir)
    if(!dir.exists()) {
      dir.mkdir()
    }
    val file = new File(dir, fileName)
    if(!file.exists())
      file.createNewFile()
    new DataOutputStream(new FileOutputStream(file))
  }

  private def writeModel(output : DataOutputStream, model: Model) : Unit = {
    try {
      val bytes = model.toBytes()
      output.writeLong(bytes.length)
      output.writeLong(model.getType)
      output.write(bytes)
    } catch {
      case t: Throwable =>
        System.out.println("Error Serializing model")
        t.printStackTrace()
    }
  }

  private def writeServingInfo(output: DataOutputStream, servingInfo: ModelToServeStats ): Unit = {
    try{
      output.writeLong(5)
      output.writeUTF(servingInfo.description)
      output.writeUTF(servingInfo.name)
      output.writeLong(servingInfo.since)
      output.writeLong(servingInfo.usage)
      output.writeDouble(servingInfo.duration)
      output.writeLong(servingInfo.min)
      output.writeLong(servingInfo.max)
    }catch {
      case t: Throwable =>
        System.out.println("Error Serializing servingInfo")
        t.printStackTrace()
    }
  }
}
